### DISCUSSION TITLE
How can I get the precision, recall, map@0.5 of the network prediction using this slicing method？

### DISCUSSION THREAD

**Discussion by Grenaf** [obss/sahi#718] [Q&A] [answered]
Hello fcakyon，in order to detect small targets, I used a lot of 640x640 slice images to train yolov5s（slice image labels are yolov5 format）, and then used sahi to take the high-resolution images synthesized for detection(high-resolution images are  prediction set, there are no labels). 
* I have a question how can I get precision, recall, map@0.5 on the prediction set or validation set(high-resolution images), my prediction set are some high-resolution pictures, and the training set are  slices of some high-resolution pictures(slice image labels are yolov5 format).
* I see a way about this question https://github.com/obss/sahi/issues/182#issuecomment-894101529, I am not sure how to use it，because my dataset labels are YOLO format。Can I use sahi to detect multiple high-resolution images at once and get its evaluation, precision, recall, map@0.5, map@0.5: 0.95？

**Comment by fcakyon** [comment]
@Grenaf thanks for your valuable feedback!

As you mentioned, you can perform sliced inference, save results in coco result format and evaluate map50 map0.5: 0.95 etc following the steps mentioned here: https://github.com/obss/sahi/issues/182#issuecomment-894101529

**Comment by fcakyon** [comment]
You have to convert your dataset into coco format before performing `sahi predict` and `sahi coco evaluate`.

**Comment by Grenaf** [comment]
> @Grenaf thanks for your valuable feedback!
> 
> As you mentioned, you can perform sliced inference, save results in coco result format and evaluate map50 map0.5: 0.95 etc following the steps mentioned here: [#182 (comment)](https://github.com/obss/sahi/issues/182#issuecomment-894101529)

![image](https://user-images.githubusercontent.com/73221711/142170590-316fff62-7aec-4278-affd-fc3a505fc87d.png)
So I need to use the COCO format label, these two parameters need to set the image path of the test set（images） and the label（COCO fomat） of the test set

**Comment by Grenaf** [comment]
> You have to convert your dataset into coco format before performing `sahi predict` and `sahi coco evaluate`.

I have a question. Two prediction methods are used in sahi. When I use the sahi coco evaluate --dataset_json_path dataset.json --result_json_path result.json command, which method will the model choose to evaluate the model。get_prediction() or get_slice_prediction()

**Comment by fcakyon** [comment]
> > You have to convert your dataset into coco format before performing `sahi predict` and `sahi coco evaluate`.
> 
> I have a question. Two prediction methods are used in sahi. When I use the sahi coco evaluate --dataset_json_path dataset.json --result_json_path result.json command, which method will the model choose to evaluate the model。get_prediction() or get_slice_prediction()

`evaluate` command does not decide that. You have to set it in the `predict` command as 
`sahi predict --no_sliced_prediction` while creating the `result.json` file.

By default, `sahi predict` performs both sliced and full inference. Check the `predict` docs for the details: https://github.com/obss/sahi/blob/main/docs/CLI.md#predict-command-usage

**Comment by fcakyon** [comment]
> > @Grenaf thanks for your valuable feedback!
> > As you mentioned, you can perform sliced inference, save results in coco result format and evaluate map50 map0.5: 0.95 etc following the steps mentioned here: [#182 (comment)](https://github.com/obss/sahi/issues/182#issuecomment-894101529)
> 
> ![image](https://user-images.githubusercontent.com/73221711/142170590-316fff62-7aec-4278-affd-fc3a505fc87d.png) So I need to use the COCO format label, these two parameters need to set the image path of the test set（images） and the label（COCO fomat） of the test set

You need 2 things:
- a coco formatted dataset json file
- and a coco formatted result json file

COCO formatted result json file can be created using `sahi predict --source image/dir/ --dataset_json_file coco_dataset.json` command.

**Comment by Grenaf** [comment]
![image](https://user-images.githubusercontent.com/73221711/142180520-a0b5293f-bfee-4774-b271-5c213b4ad7ee.png)
So I need these files, annotation file and results.json?

**Comment by fcakyon** [comment]
> ![image](https://user-images.githubusercontent.com/73221711/142180520-a0b5293f-bfee-4774-b271-5c213b4ad7ee.png) So I need these files, annotation file and results.json?

You are right!

**Comment by fcakyon** [comment]
@Grenaf closing this issue if you dont have any more questions :)

**Comment by Grenaf** [comment]
# Results do not correspond to current coco set
* sahi predict --model_confidence_threshold 0.6 --source F:\deeplearning_project\sahi_0.8.13\FWX\original\original_test_imageslabels --model_path F:\deeplearning_project\sahi_0.8.13\FWX\original\original_best.pt --model_type yolov5 --no_sliced_prediction --dataset_json_path F:\deeplearning_project\sahi_0.8.13\FWX\original\original_test_imageslabels.json --export_visual
Then i get result.json
* sahi coco evaluate --dataset_json_path F:\deeplearning_project\sahi_0.8.13\FWX\original\original_test_imageslabels.json --result_json_path F:\deeplearning_project\sahi_0.8.13\FWX\original\result.json
but an error occured
![image](https://user-images.githubusercontent.com/73221711/145678503-71ebdbdc-cfb9-4f29-b25e-3a55f6fa53e9.png)


I used these two instructions, but it prompted an error. I'm not sure what the problem is.

**Comment by fcakyon** [comment]
@Grenaf this issue is fixed with: https://github.com/obss/sahi/pull/299 but havent released a new version for sahi yet. You can either install sahi from main branch with: `pip install git+https://github.com/obss/sahi.git@main` or wait for a new release (will happen today :100: ).

**Comment by Grenaf** [comment]
> this issue is fixed with: #299 but havent released a new version for sahi yet. You can either install sahi from main branch with: `pip install git+https://github.com/obss/sahi.git@main` or wait for a new release (will will happen today).

Okay, don’t worry, look forward to your new news.

**Comment by fcakyon** [comment]
@Grenaf latest version (0.8.14) is live!

**Comment by Grenaf** [comment]
> @Grenaf latest version (0.8.14) is live!
* Today I used sahi 0.8.14, but a new problem has arisen, and I am a little confused about the results of the evaluation.
  *  I use sahi predict --model_confidence_threshold 0.6 --source FWX/original/original_test_imageslabels --model_path FWX/original/original_best.pt --model_type yolov5 --no_sliced_prediction --dataset_json_path FWX/original/original_test_imageslabels.json --export_visual
  * sahi coco evaluate --dataset_json_path FWX/original/original_test_imageslabels.json --result_json_path FWX/original/result.json --out_dir FWX/output/coco --type bbox --iou_thrs 0.5

the result just like this，Is this result normal? Through the visuals_with_gt obtained by the first command, I saw that the model successfully predicted the target. I think the values of map and ap may be abnormal。
![image](https://user-images.githubusercontent.com/73221711/145706110-05839e16-14bb-47cb-8f6c-edc32b07370c.png)

**Comment by fcakyon** [comment]
@Grenaf it means that your category ids in your trained model and coco json dataset does not match. You have to fix the `categories` field of your coco dataset json to have proper evaluation.

In most cases, mapping category ids from 0 to N solves the issue. If you can provide your model i can tell you the proper category id to name mapping.

**Comment by Grenaf** [comment]
* I use the yolov5s model, there are a total of six categories of targets to be detected. The original yolov5s label category id is 0-5, in order to use coco evaluate I need to change the yolo label to coco format https://github.com/Taeyoung96/Yolo-to-COCO-format-converter, and then the category id becomes 1-6. So what documents should I provide? Thanks

**Comment by Grenaf** [comment]
> @Grenaf it means that your category ids in your trained model and coco json dataset does not match. You have to fix the `categories` field of your coco dataset json to have proper evaluation.
> 
> In most cases, mapping category ids from 0 to N solves the issue. If you can provide your model i can tell you the proper category id to name mapping.
* I found the problem. When I convert the yolo label to the coco format, the target category id in original_test_imageslabels.json is 1-6
![image](https://user-images.githubusercontent.com/73221711/145792140-0b808a83-c74b-4736-9769-735e253ab295.png)

* When I execute sahi predict --model_confidence_threshold 0.6 --source FWX/original/original_test_imageslabels --model_path FWX/original/original_best.pt --model_type yolov5 --no_sliced_prediction --dataset_json
_path FWX/original/original_test_imageslabels.json --export_visual
* result.json ，the category in result.json does not match the category in original_test_imageslabels.json. Taking missing categories id as an example, its id in original_test_imageslabels.json is 1, but the id in result is 0. So the problem is categories id Caused by a mismatch. So is there a way to solve this problem，do I need to modify the category information in the original_test_imageslabels.json file?Looking forward to your reply！
![image](https://user-images.githubusercontent.com/73221711/145792276-cd934613-0873-4d5b-9f1b-e4229a958f4f.png)

**Comment by fcakyon** [comment, marked as answer]
You model is trained with category ids 0-5 but your dataset json contains category ids 1-6. You have to map your category ids in the dataset json to 0-5 to properly perform evaluation. You can use sahi coco utilities to manage that: https://github.com/obss/sahi/blob/main/docs/COCO.md

**Comment by IrDIE** [comment]
I implemented validation with SAHI for YOLO8
https://github.com/IrDIE/YOLO8_SAHI/tree/master

💙 If you like it - press ⭐
